diff --git a/go.mod b/go.mod
index e7f308e890..def9507093 100644
--- a/go.mod
+++ b/go.mod
@@ -5,6 +5,31 @@ go 1.21
 // Client-go does not handle different versions of mergo due to some breaking changes - use the matching version
 replace github.com/imdario/mergo => github.com/imdario/mergo v0.3.5

+require (
+	mosn.io/htnn/controller v0.0.0
+	mosn.io/htnn/types v0.2.1
+)
+
+replace (
+	mosn.io/htnn/api => ../htnn/api
+	mosn.io/htnn/controller => ../htnn/controller
+	mosn.io/htnn/types => ../htnn/types
+)
+
+// Downgrade some deps to match the one used in istio
+replace (
+	github.com/prometheus/common => github.com/prometheus/common v0.45.0
+	github.com/prometheus/client_golang => github.com/prometheus/client_golang v1.18.0
+	github.com/envoyproxy/go-control-plane => github.com/envoyproxy/go-control-plane v0.12.1-0.20240326194405-485b2263e153
+)
+
+// The break change https://github.com/google/cel-go/commit/04536922515e3de9de498f689b40f70d170b518b
+// added in v0.18.0 causes `go test istio.io/istio/pkg/config/crd` to fail because of compilation error.
+// We can't update the test to solve the error because it's indirect depended by k8s/apiserver.
+replace github.com/google/cel-go => github.com/google/cel-go v0.17.8
+// HTNN requires a higher version of opentelemetry package
+replace go.opentelemetry.io/otel/sdk/metric => go.opentelemetry.io/otel/sdk/metric v1.28.0
+
 require (
 	cloud.google.com/go/compute/metadata v0.2.3
 	cloud.google.com/go/logging v1.9.0
diff --git a/pilot/pkg/config/htnn/types.go b/pilot/pkg/config/htnn/types.go
new file mode 100644
index 0000000000..8d39ede5b0
--- /dev/null
+++ b/pilot/pkg/config/htnn/types.go
@@ -0,0 +1,6 @@
+package htnn
+
+import (
+   _ "mosn.io/htnn/types/plugins"    // register plugin types
+   _ "mosn.io/htnn/types/registries" // register registry types
+)
