diff --git a/pkg/istio-agent/xds_proxy_delta.go b/pkg/istio-agent/xds_proxy_delta.go
index bf0603182369..31c2f092c8bd 100644
--- a/pkg/istio-agent/xds_proxy_delta.go
+++ b/pkg/istio-agent/xds_proxy_delta.go
@@ -288,13 +288,9 @@ func (p *XdsProxy) deltaRewriteAndForward(con *ProxyConnection, resp *discovery.
 		return
 	}

-	respResources := make([]*discovery.Resource, 0, len(resources))
 	for i := range resources {
-		respResources = append(respResources, &discovery.Resource{
-			Resource: resources[i],
-		})
+		resp.Resources[i].Resource = resources[i]
 	}
-	resp.Resources = respResources

 	proxyLog.Debugf("forward ECDS resources %+v", resp.Resources)
 	forward(resp)
